### PR TITLE
Adapt JSON serialization

### PR DESCRIPTION
**Pull request by danielfett**
This implements JSON serialization as outlined in https://github.com/oauth-wg/oauth-selective-disclosure-jwt/issues/403 and defined in https://github.com/oauth-wg/oauth-selective-disclosure-jwt/pull/414

Noteworthy changes:

- To support JSON Serialization with multiple issuer signatures:
  - In the settings.yml files for examples and test cases, issuer_key was renamed to issuer_keys and MAY be an array of keys.
  - The `issuer_keys` argument to the SDJWTIssuer MAY be an array of keys (but a single key is still supported, so no changes in legacy usages required)
  - The callable `cb_get_issuer_key` passed to SDJWTVerifier MAY return a JWKSet
 - Example/test case specification.yml files may now specify `settings_override` to override whatever is defined in the `settings.yml`

### CODE REVIEW

File: src/sd_jwt/bin/demo.py [hunk 1]
```diff
@@ -156,82 +164,92 @@ def run():
...
+            else None
+        ),
         _args.nonce if example.get("key_binding", False) else None,
+        serialization_format=serialization_format,
     )
     verified = sdjwt_at_verifier.get_verified_payload()
 
     ### Done - now output everything to CLI (unless --replace-examples-in was used)
 
-    iid_payload = ""
```
Comment (danielfett) [inline src/sd_jwt/bin/demo.py:200]: Note: demo.py was broken, changes in this file are just to revive it.

**Comment by bc-pi** [comment]
> * `issuer_keys` argument to the SDJWTIssuer MAY be an array of keys (but a single key is still supported, so no changes in legacy usages required)

but an update to the name in the settings.yml of dependent projects (like [SD-JWT VC](https://github.com/oauth-wg/oauth-sd-jwt-vc/blob/f3e8ceb7637f351112dc02d9a9ea274f8a7cf040/examples/settings.yml#L10)) will be needed, no?

File: examples/settings.yml [hunk 2]
```diff
@@ -7,12 +7,12 @@ key_settings:
 
     kty: EC
 
-    issuer_key:
-        kty: EC
-        d: Ur2bNKuBPOrAaxsRnbSH6hIhmNTxSGXshDSUD1a1y7g
-        crv: P-256
-        x: b28d4MwZMjw8-00CG4xfnn9SLMVMM19SlqZpVb_uNtQ
-        y: Xv5zWwuoaTgdS6hV43yI6gBwTnjukmFQQnJ_kCxzqk8
+    issuer_keys:
```
Comment (peppelinux) [inline examples/settings.yml:10]: yes, issuer_keys is supposed to be an ARRAY even if single valued. agree.

File: src/sd_jwt/common.py [hunk 3]
```diff
@@ -122,30 +124,81 @@ def _check_for_sd_claim(self, the_object):
             return
 
     def _parse_sd_jwt(self, sd_jwt):
-
         if self._serialization_format == "compact":
             (
                 self._unverified_input_sd_jwt,
                 *self._input_disclosures,
-                self._unverified_input_key_binding_jwt
+                self._unverified_input_key_binding_jwt,
```
Comment (peppelinux) [inline src/sd_jwt/common.py:131]:
```suggestion
                self._unverified_input_key_binding_jwt
```

File: src/sd_jwt/common.py [hunk 4]
```diff
@@ -122,30 +124,81 @@ def _check_for_sd_claim(self, the_object):
...
+                self._unverified_input_key_binding_jwt = (
+                    self._unverified_input_sd_jwt_parsed["header"].get(
+                        JSON_SER_KB_JWT_KEY, ""
+                    )
+                )
+                self._unverified_compact_serialized_input_sd_jwt = ".".join(
+                    [
+                        self._unverified_input_sd_jwt_parsed["protected"],
+                        self._unverified_input_sd_jwt_parsed["payload"],
+                        self._unverified_input_sd_jwt_parsed["signature"],
```
Comment (peppelinux) [inline src/sd_jwt/common.py:167]:
```suggestion
                        self._unverified_input_sd_jwt_parsed["signature"]
```

File: src/sd_jwt/common.py [hunk 5]
```diff
@@ -122,30 +124,81 @@ def _check_for_sd_claim(self, the_object):
...
+            else:
+                raise ValueError("Invalid JSON serialization of SD-JWT")
+
+    def _calculate_kb_hash(self, disclosures):
+        # Temporarily create the combined presentation in order to create the hash over it
+        # Note: For JSON Serialization, the compact representation of the SD-JWT is restored from the parsed JSON (see common.py)
+        string_to_hash = self._combine(
+            self._unverified_compact_serialized_input_sd_jwt,
+            *disclosures,
+            "",
```
Comment (peppelinux) [inline src/sd_jwt/common.py:202]:
```suggestion
            ""
```

File: src/sd_jwt/utils/demo_utils.py [hunk 6]
```diff
@@ -54,17 +54,23 @@ def get_jwk(jwk_kwargs: dict = {}, no_randomness: bool = False, random_seed: int
...
+        issuer_keys = [JWK.generate(**_kwargs)]
         holder_key = JWK.generate(**_kwargs)
 
-    issuer_public_key = JWK.from_json(issuer_key.export_public())
+    if len(issuer_keys) > 1:
+        issuer_public_keys = JWKSet()
+        for k in issuer_keys:
+            issuer_public_keys.add(JWK.from_json(k.export_public()))
+    else:
+        issuer_public_keys = JWK.from_json(issuer_keys[0].export_public())
```
Comment (peppelinux) [inline src/sd_jwt/utils/demo_utils.py:70]: I would normalize it to array anyway, and therefore let the code flow ... go

```suggestion
    if instance(issuer_keys, dict):
        issuer_keys = [issuer_keys]
    issuer_public_keys = JWKSet()
    for k in issuer_keys:
            issuer_public_keys.add(JWK.from_json(k.export_public()))
```

File: tests/test_disclose_all_shortcut.py [hunk 7]
```diff
@@ -61,9 +64,13 @@ def cb_get_issuer_key(issuer, header_parameters):
...
     assert verified == expected_claims
 
+    # We don't compare header parameters for JSON Serialization for now
+    if serialization_format != "compact":
+        return
+
     expected_header_parameters = {
         "alg": testcase.get("sign_alg", "ES256"),
-        "typ": "testcase+sd-jwt"
+        "typ": "testcase+sd-jwt",
```
Comment (peppelinux) [inline tests/test_disclose_all_shortcut.py:73]:
```suggestion
        "typ": "testcase+sd-jwt"
```

**Comment by c2bo** [comment]
> > * `issuer_keys` argument to the SDJWTIssuer MAY be an array of keys (but a single key is still supported, so no changes in legacy usages required)
> 
> but an update to the name in the settings.yml of dependent projects (like [SD-JWT VC](https://github.com/oauth-wg/oauth-sd-jwt-vc/blob/f3e8ceb7637f351112dc02d9a9ea274f8a7cf040/examples/settings.yml#L10)) will be needed, no?

What about catching the old yaml files during parsing, converting to the correct claim names (basically issuer_key -> issuer_keys for both functions parsing yaml files) and outputting a warning that `issuer_key` is deprecated? That would be a very small change and not break stuff.

File: src/sd_jwt/utils/demo_utils.py (hunk 6 above)
Comment (danielfett) [reply src/sd_jwt/utils/demo_utils.py:70]: Let's refactor later!

**Pull request merged**